docs/adr/ADR-023-sequencer-recovery.md

```diff
@@ -0,0 +1,132 @@
+### User requirements
+- **No split‑brain**: at most one sequencer is active.
+- **Deterministic recovery**: new leader starts from a known **unsafe head**.
+- **Fast failover**: p50 ≤ 15s, p95 ≤ 45s.
+- **Operational clarity**: health metrics, leader identity, and explicit admin controls.
+- **Zero‑downtime upgrades**: blue/green leadership transfer.
+
+### Systems affected
+- `ev-node` (sequencer control hooks, health surface).
+- New sidecar(s): **conductor** (Design A) or **lease‑manager** (Design B).
+- RPC ingress (optional **leader‑aware proxy** to route sequencing endpoints only to the leader).
+- CI/CD & SRE runbooks, dashboards, alerts.
+
+### New/changed data structures
+- **UnsafeHead** record persisted by control plane: `(l2_number, l2_hash, l1_origin, timestamp)`.
+- **Design A (Raft)**: replicated **Raft log** entries for `UnsafeHead`, `LeadershipTerm`, and optional `CommitMeta` (batch/DA pointers); periodic snapshots.
+- **Design B (Lease)**: a single **Lease** record (Kubernetes Lease or external KV entry) plus a monotonic **lease token** for fencing.
+
+### New/changed APIs
+Introduce an **Admin RPC** (gRPC/HTTP) on `ev-node` (or a thin shim) used by either control plane:
+
+- `StartSequencer(from_unsafe_head: bool)` — start sequencing, optionally pinning to the last persisted UnsafeHead.
+- `StopSequencer()` — hard stop; no more block production.
+- `SequencerHealthy()` → `{ healthy, l2_number, l2_hash, l1_origin, peer_count, da_height, last_err }`
+- `Status()` → `{ sequencer_active, build_height, leader_hint?, last_err }`
+
+These are additive and should not break existing RPCs.
+
+### Efficiency considerations
+- **Design A:** Raft heartbeats and snapshotting add small steady‑state overhead; no impact on throughput when healthy.
+- **Design B:** Lease renewals are lightweight; performance dominated by `ev-node` itself.
+
+### Expected access patterns
+- Reads (RPC, state) should work on all nodes; **writes/sequence endpoints** only on the active leader. If a leader‑aware proxy is deployed, it enforces this automatically.
+
+### Logging/Monitoring/Observability
+- Metrics: `leader_id`, `raft_term` (A), `lease_owner` (B), `unsafe_head_advance`, `peer_count`, `rpc_error_rate`, `da_publish_latency`, `backlog`.
+- Alerts: no unsafe advance > 3× block time; unexpected leader churn; lease lost but sequencer still active (fencing breach).
+- Logs: audit all **Start/Stop** decisions and override operations.
+
+### Security considerations
+- Lock down **Admin RPC** with mTLS + RBAC; only the sidecar/process account may call Start/Stop.
+- Implement **fencing**: leader periodically validates it still holds leadership/lease; otherwise self‑stops.
+- Break‑glass overrides must be gated behind separate credentials and produce auditable events.
+
+### Privacy considerations
+- None beyond existing node telemetry; no user data added.
+
+### Testing plan
+- Kill active sequencer → verify failover within SLO; assert **no double leadership**.
```

Comment:
For Design A, we should also kill the conductor on the active sequencer so that others conductors can experience a timeout from the conductor leader.